lib/encoding/yaml/parser.toit

```diff
@@ -0,0 +1,1603 @@
+
+  ns-yaml-version -> bool:
+    mark := mark
+    return try-parse --as-bool:
+      if (repeat --at-least-one: ns-dec-digit) and
+          match-char '.' and
+         (repeat --at-least-one: ns-dec-digit):
+        version := string-since mark
+        parts := version.split "."
+        major := int.parse parts[0]
+        minor := int.parse parts[1]
+        if major > 1 or minor > 2:
+          set-error "UNSUPPORTED_YAML_VERSION"
+          false
+        else:
+          true
+
+  ns-tag-directive -> bool:
+   return try-parse --as-bool:
+     match-string "TAG" and
+         s-separate-in-line and
+         c-tag-handle and
+         s-separate-in-line and
+         ns-tag-prefix
+
+  c-tag-handle -> bool:
+    return c-named-tag-handle or match-string S-SECONDARY-TAG-HANDLE_ or (match-char C-TAG_) != null
+
+  c-named-tag-handle -> bool:
+    return try-parse --as-bool:
+      match-char C-TAG_ and
+          (repeat --at-least-one : ns-word-char) and
+          match-char C-TAG_
+
+  ns-tag-prefix -> bool:
+    return c-ns-local-tag-prefix or ns-global-tag-prefix
+
+  c-ns-local-tag-prefix -> bool:
+    return try-parse --as-bool: match-char C-TAG_ and (repeat: ns-uri-char)
+
+  ns-global-tag-prefix -> bool:
+    return try-parse --as-bool: ns-tag-char and (repeat: ns-uri-char)
+
+  ns-reserved-directive -> bool:
+    return try-parse --as-bool: ns-directive-name and (repeat: s-separate-in-line and ns-directive-parameter)
+
+  ns-directive-name -> List?:
+    return repeat --at-least-one: ns-char
+
+  ns-directive-parameter -> List?:
+    return repeat --at-least-one: ns-char
+
+  // Comments.
+  l-comment -> bool:
+    return try-parse --as-bool: s-separate-in-line and (optional: c-nb-comment-text) and b-comment
+
+  s-l-comments -> bool:
+    return try-parse --as-bool: (s-b-comment or start-of-line) and (repeat: l-comment)
+
+  s-b-comment -> bool:
+    return try-parse --as-bool: (optional: s-separate-in-line and (optional: c-nb-comment-text)) and b-comment
+
+  s-separate-in-line -> bool:
+    return try-parse --as-bool: (repeat --at-least-one: s-white) or start-of-line
+
+  c-nb-comment-text -> bool:
+    return as-bool (c-comment and (repeat: nb-char))
+
+  b-comment -> bool:
+    return b-non-content or l-eof
+
+  b-non-content -> bool:
+    return b-break
+
+  l-trail-comments n/int -> bool:
+    return try-parse --as-bool:
+      s-indent-less-than n and
+          c-nb-comment-text and
+          b-comment and
+          (repeat: l-comment)
+
+  // Data part.
+  s-l-plus-block-node n/int c/int -> ValueNode_?:
+    return s-l-plus-block-in-block n c or s-l-plus-flow-in-block n
+
+  s-l-plus-block-in-block n/int c/int -> ValueNode_?:
+    return s-l-plus-block-scalar n c or s-l-plus-block-collection n c
+
+  s-l-plus-flow-in-block n/int -> ValueNode_?:
+    if node := (try-parse: s-separate n + 1 FLOW-OUT_ and ns-flow-node n + 1 FLOW-OUT_):
+      if s-l-comments: return node
```

Comment:
I might be off, but here we parse the `node`, but if the `s-l-comments` is false, we don't roll back.

Reply:
Good catch. Added a test case as guard.